keps/0007-dynamic-audit-control.md

```diff
@@ -0,0 +1,126 @@
+
+### Goals
+- Provide an api and set of objects to configure the advanced auditing kube-apiserver webhook flags dynamically
+- Allow the current audit policy object to be configured dynamically
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration of Policy
+Dynamic configuration of the existing policy object will be provided through use of a shared informer.
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: Policy
+omitStages:
+  - "RequestReceived"
+rules:
+  - level: <request level>
+    resources:
+    - group: <group>
+      resources: <resources>
+```
+In kube-apiserver today this is only accepted as a file, this KEP would allow it to either be configured as a file on the master node or applied to the `kube-system` namespace as an object and be dynamically configured. This will require the audit API to be registered with the apiserver. If both file and object exist the dynamic object will take precedence.
+
+### Dynamic Configuration of Flags
+There are a number of configuration options that are currently set through runtime flags. This KEP would provide a new API for the dynamic configuration of the webhook parameters.
+```yaml
+apiVersion: audit.k8s.io/v1alpha1
+kind: AuditConfiguration
+metadata:
+  name: <name>
+webhook:
+  mode: <batch or blocking>
+  initialBackoff: <10s>
+  batch:
+    bufferSize: <400>
+    maxSize: <10000>
+    maxWait: <30s>
+    throttleBurst: <15>
+    throttleEnabled: <true>
+    throttleQPS: <10>
+```
+The webhook would share the model of validating and mutating admission control plugins, leveraging a shared informer to provide online configuration. If runtime flags are also provided, the dynamic configuration will override any parameters provided there.
+
+### User Stories
+
+#### Story 1
+As a cluster admin I will easily be able to enable the interal auditing features of an existing cluster, and tweak the configurations as necessary.
+
+#### Story 2
+As a Kubernetes extension developer, I will be able to provide drop in extensions that utilize audit data.
+
+### Implementation Details/Notes/Constraints
+
+A misunderstanding may occur of both flags are defined and an object is provided. This should be explicitly called out in the 
+documentation and monitored within the community. If this becomes a point of contention we will need to reevaluate the implementation.
+
+### Risks and Mitigations
+
+This does open up the attack surface of the audit mechanisms. Having them strictly configured through the api server has 
+the advantage of limiting the access of those configurations to those that have access to the master node. However, Dynamic Admission has shown that this can be done in a safe manner. 
```

Comment:
> As a potential mitigation strategy, what if we said that the flag-defined configuration always takes precedence, but can be combined with the dynamic config? This would let the cluster operator define a policy subset that must be abided by (e.g. ALWAYS log secrets access, at metadata level; NEVER log events, ALWAYS log dynamic audit log changes). This could easily be done by just appending the dynamic config to the static config.

that's tricky... I can see valid use cases both ways:

* flag-defined config that limits what dynamic configs can do
* flag-defined config for operating in normal mode, dynamic config that allows temporary verbose logging to debug a particular issue

Reply:
good idea 👍 everyone wins with this, updated

Reply:
Temporary verbose logging is an interesting use case. Could we settle on assuming that if you want that feature you would need to leave your static config open to it, and configure the rest dynamically?

Reply:
Having dynamic audit config overriding flag-defined config as defined is very problematic, IMHO. Dynamic policies, e.g. defined as an API object, would be letting the API monitoring itself. It would remove the ability to provide any guarantee on the integrity of an audit.

The only tenable way to allow some dynamism to audit policy would be to have flag-defined config be the ultimate authority and let it specify the bounds and policy for applying any dynamic configuration.

Reply:
> As a potential mitigation strategy, what if we said that the flag-defined configuration always takes precedence, but can be combined with the dynamic config? This would let the cluster operator define a policy subset that must be abided by (e.g. ALWAYS log secrets access, at metadata level; NEVER log events, ALWAYS log dynamic audit log changes). This could easily be done by just appending the dynamic config to the static config.

Agree to make the flag-defined configuration read only.
And the dynamic audit config will send audits to it's own backends, they should not effect each other.